jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/RequestTest.java

```diff
@@ -548,73 +546,394 @@ private static void checkCookieResult(String containedCookie, String[] notContai
+        cases.add(Arguments.of("%E5%B8%BD%E5%AD%9=Beret&other=foo", 400, "", "")); // different from 11
+        cases.add(Arguments.of("%E5%B8%BD%E5%AD%=Beret&other=foo", 400, "帽子", ""));
+        cases.add(Arguments.of("%E5%B8%BD%E5%AD=Beret&other=foo", 200, "帽�", "Beret"));
+
+        // raw unicode parameter names (strange replacement logic here)
+        cases.add(Arguments.of("€=currency", 200, "?", "currency"));
+        cases.add(Arguments.of("帽子=Beret", 200, "??", "Beret"));
+
+        // invalid pct-encoded parameter name
+        cases.add(Arguments.of("foo%xx=abc", 400, "foo�", ""));
+        cases.add(Arguments.of("foo%x=abc", 400, "foo�", ""));
+        cases.add(Arguments.of("foo%=abc", 400, "foo�", ""));
+
+        // utf-16 values (LEGACY has UTF16_ENCODINGS enabled, but it doesn't work for query apparently)
+        cases.add(Arguments.of("foo=a%u2192z", 400, "foo", ""));
+
+        // truncated utf-16 values (LEGACY has UTF16_ENCODINGS enabled, but it doesn't work for query apparently)
+        cases.add(Arguments.of("foo=a%u219z", 400, "foo", ""));
+
+        return cases.stream();
+    }
+
     @ParameterizedTest
-    @ValueSource(strings = { "true", "false"})
-    public void testBadUtf8Query(boolean allowBadUtf8) throws Exception
+    @MethodSource("queryBehaviorsLegacy")
+    public void testQueryExtractionBehaviorLegacy(String inputQuery, int expectedStatus, String expectedKey, String expectedValue) throws Exception
     {
-        server.stop();
-        List<ComplianceViolation.Event> events = new CopyOnWriteArrayList<>();
-        ComplianceViolation.Listener listener = new ComplianceViolation.Listener()
-        {
-            @Override
-            public void onComplianceViolation(ComplianceViolation.Event event)
-            {
-                events.add(event);
-            }
-        };
+        UriCompliance uriCompliance = UriCompliance.LEGACY;
+        testQueryExtractionBehavior(uriCompliance, inputQuery, expectedStatus, expectedKey, expectedValue);
+    }
 
-        if (allowBadUtf8)
-        {
-            for (Connector connector : server.getConnectors())
-            {
-                HttpConnectionFactory httpConnectionFactory = connector.getConnectionFactory(HttpConnectionFactory.class);
-                if (httpConnectionFactory != null)
-                {
-                    HttpConfiguration httpConfiguration = httpConnectionFactory.getHttpConfiguration();
-                    httpConfiguration.setUriCompliance(UriCompliance.DEFAULT.with("test", UriCompliance.Violation.BAD_UTF8_ENCODING));
-                    httpConfiguration.addComplianceViolationListener(listener);
-                }
-            }
-        }
+    /**
+     * Behaviors as they existed in Jetty 12.0.16
+     */
+    public static Stream<Arguments> queryBehaviorsDefault()
+    {
+        List<Arguments> cases = new ArrayList<>();
+
+        // TODO: All of these 500 response status codes should really be 400
```

Comment:
This line can be deleted now that there are no 500 responses in this list of cases.